### ISSUE TITLE
bos, mask embedding

### ISSUE THREAD

**Issue by taiseitosaki** [aratake-10/250111_Trans#2]
https://github.com/aratake-10/250111_Trans/blob/0a797539e759f26cceb4143c2b039b08afa70157/src/column.py#L37
https://github.com/aratake-10/250111_Trans/blob/0a797539e759f26cceb4143c2b039b08afa70157/src/column.py#L38
これは良くないのでは？
bos単独のembedding、mask単独のembeddingにならず、[bos, mask] sentenceのbos embeddingとかmask embeddingになってます。bidirectional なので、maskからのattentionがbosに向かっちゃってますよね？

**Comment by aratake-10** [comment]
ありがとうございます。誤って[bos, mask]というシーケンスを処理してしまっていたところを、明示的に[bos]と[mask]それぞれの埋め込み表現を取得するように確認しました。

のちほどまた全体の修正版をpushする予定ですが、取り急ぎ問題修正後のcolumn.pyです。

```
import torch
import numpy as np
import pandas as pd
from sklearn.decomposition import PCA
from transformers import BertModel, BertTokenizer
from typing import Optional
from scipy.io import arff
from utils import set_seed, set_device

set_seed(42)

# BERT埋め込みを取得するためのモデルクラス
class BertEmbeddings(torch.nn.Module):
    def __init__(self, model_name='bert-base-uncased'):
        super().__init__()
        self.bert = BertModel.from_pretrained(model_name)
        self.tokenizer = BertTokenizer.from_pretrained(model_name)  # Tokenizerを追加
        

    def forward(self, input_ids: Optional[torch.LongTensor] = None) -> torch.Tensor:
        if input_ids is not None:
            outputs = self.bert(input_ids)
            inputs_embeds = outputs.last_hidden_state
        else:
            raise ValueError("input_ids must be provided")


        return inputs_embeds

# MASKとBOSトークン埋め込みを取得する関数:ここを1/11修正
def get_mask_and_bos_embeddings(model, bos_token_id=101, mask_token_id=103):
    # [BOS]tokenのembed取得
    bos_input_ids = torch.tensor([[bos_token_id]])
    bos_embedding = model(bos_input_ids)
    
    # [MASK]tokenのembed取得
    mask_input_ids = torch.tensor([[mask_token_id]])
    mask_embedding = model(mask_input_ids)
    
    #このままだと(1, 1, 768)なので(768)にshapeを合わせる。
    bos_embedding = bos_embedding[0, 0, :]
    mask_embedding = mask_embedding[0, 0, :]
    
    return bos_embedding, mask_embedding # それぞれのshapeはこの時点で(1,1,768) 

# PCAによる次元圧縮
def apply_pca(embeddings, n_components=8, seed=42):
    # シード固定
    np.random.seed(seed)
    torch.manual_seed(seed)
    
    pca = PCA(n_components=n_components)
    compressed_embeddings = pca.fit_transform(embeddings.detach().cpu().numpy())  # detachを追加
    return torch.tensor(compressed_embeddings, dtype=torch.float)

# process_data関数内に追加
def process_data(file_path, model, n_components=8):
    # ARFFファイルの読み込み
    data, meta = arff.loadarff(file_path)
    df = pd.DataFrame(data)

    # 列名ごとにBERT埋め込みを取得
    embeddings = []
    for column_name in df.columns:
        # 列名をトークン化
        input_ids = torch.tensor([model.tokenizer.encode(column_name, add_special_tokens=True)])  # 列名をトークン化
        column_embedding = model(input_ids)  # BERT埋め込み取得
        embeddings.append(column_embedding[0, 0, :])  # 最初のトークン（[CLS]トークン）の埋め込み

    embeddings_tensor = torch.stack(embeddings)

    # MASKとBOS埋め込みを取得
    mask_embedding, bos_embedding = get_mask_and_bos_embeddings(model)

    # すべての埋め込みを結合（BOSとMASKを含めて11次元）
    all_embeddings = torch.cat([bos_embedding.unsqueeze(0), embeddings_tensor, mask_embedding.unsqueeze(0)], dim=0)  # (11, 768)

    # PCA適用（11次元を8次元に圧縮）
    reduced_embeddings = apply_pca(all_embeddings, n_components=n_components)  # (11, 8)

    # MASK埋め込みを取り除き、BOSを最初に持ってきて、最終的な形に
    final_embeddings = torch.cat([reduced_embeddings[0].unsqueeze(0), reduced_embeddings[1:-1]], dim=0)

    # final_embeddingsの形状を(全サンプル数, 列数, d_embedding)に拡張
    final_embeddings = final_embeddings.unsqueeze(0).repeat(df.shape[0], 1, 1)  # [全サンプル数, 列数, 8]

    # PCA後の8次元のMASK埋め込みを保存(MASKは一番後ろの列にある)
    torch.save(reduced_embeddings[-1], '../data/bert_mask_embed.pt')

    # 最終的な10次元の埋め込みを返す
    return final_embeddings



# ColumnEmbedPreprocessorクラス
class ColumnEmbedPreprocessor:
    def __init__(self, processed_file_path, d_embedding, output_dir, device, config):
        self.processed_file_path = processed_file_path
        self.d_embedding = d_embedding
        self.output_dir = output_dir
        self.device = device
        self.config = config

        # BERTモデルの準備
        self.model = BertEmbeddings()

    def process_data(self):
        # ARFFファイルのデータ処理
        return process_data(self.processed_file_path, self.model, n_components=self.d_embedding)
```

**Comment by aratake-10** [comment]
おいおいn_componentsやmodel_nameもconfigから呼び出せるようにしようと思います。
